subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/ModuleResolveState.java

```diff
@@ -310,4 +322,27 @@ void addPendingNode(NodeState node) {
         pendingDependencies.addNode(node);
     }
 
+
+    public boolean maybeUpdateSelection() {
+        ComponentState newSelected = selectorStateResolver.selectBest(getId(), getSelectors());
+        if (selected == null) {
+            select(newSelected);
+            return true;
+        } else if (newSelected != selected) {
+            changeSelection(newSelected);
+            return true;
+        }
+        return false;
+    }
+
+    boolean hasCompetingForceSelectors() {
+        if (selectors.size() > 1) {
+            for (SelectorState selector : selectors) {
+                if (selector.isForce() && !selector.getRequested().matchesStrictly(selected.getComponentId())) {
+                    return true;
```

Comment:
:o: Wouldn't this return true if we have one edge that forces, and the other that doesn't?

Reply:
If one edge forces, it is selecting the version since we short circuit on a force, so the second part of the test would fail. Or are you thinking about a special case?

Reply:
So we're doing this because it's possible for 2 leaves to have 2 different forces. But, say we first see a leaf that is not forced, on [1.0,1.1]. Then we see a leaf on forced 1.0. Wouldn't we have 2 selectors, there? One forced, the other not?